Code review2

IlyaChukanov opened the pull request.

IlyaChukanov committed:
first commit

IlyaChukanov committed:
Update hash_map.h

Change

IlyaChukanov committed:
Update hash_map.h

https://official.contest.yandex.ru/contest/17022/run-report/31325819/

IlyaChukanov committed:
Update hash_map.h

IlyaChukanov committed:
Update hash_map.h

https://official.contest.yandex.ru/contest/17022/problems/

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+    void clear();
+
+    void DoubleSize();
```
эта функция не должна быть публичной

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+template <class KeyType, class ValueType, class Hash>
+auto HashMap<KeyType, ValueType, Hash>::end() const -> const_iterator 
+{
```
эта скобка вполне помещается на предыдущей строке

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+ private:
+    const size_t initialTableSize_ = 20;
+
+    size_t hash_table_size_quantity_elements_;
```
тут тавтология в названии, оставьте просто hash_table_size_

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+    std::list<KeyValuePair> all_elements_;
+
+    Hash function_hash_number_;
```
назовите просто hash_function_

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+    ValueType& operator[](const KeyType key);
+
+    const ValueType& at(const KeyType key) const;;
```
вы не забываете отправлять код в Яндекс Контест для самопроверки? не уверен, что система одобрит 2 точки с запятой

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+template <class KeyType, class ValueType, class Hash>
+void HashMap<KeyType, ValueType, Hash>::insert(
+    std::pair<KeyType, ValueType> key_value_pair) {
```
изучите https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions и сделайте по всему коду правильные отступы: у вас в теле функции отступ четыре пробела, значит тут нужно сделать 8, чтобы объявление не сливалось с телом

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+    bool empty() const;
+
+    void insert(std::pair<KeyType, ValueType> key_value_pair);
```
можно использовать KeyValuePair

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+    HashMap(Hash hasher_obj = Hash());
+    
+    HashMap(std::initializer_list<std::pair<KeyType, ValueType>> list,
```
можно использовать KeyValuePair

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+    if (hash_table_size_quantity_elements_ > num_hash_table_elements_ * 2) {
+        return;
+    }
+    std::list<std::pair<KeyType, ValueType>> position_of_scale;
```
Нет, так не пойдёт, вы тут удваиваете хранимые в структуре объекты. Это становится очень критично, если KeyType и ValueType - большие структуры. Перепишите этот метод так, чтобы этого копирования не было.

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+template <class KeyType, class ValueType, class Hash>
+const ValueType& HashMap<KeyType, ValueType, Hash>::at(
+    KeyType key) const {
```
нужен удвоенный отступ, здесь, и в аналогичном месте в функции ниже

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+
+template <class KeyType, class ValueType, class Hash>
+HashMap<KeyType, ValueType, Hash>& HashMap<KeyType, ValueType, Hash>::operator=(
+    HashMap<KeyType, ValueType, Hash> const& other) {
```
нужен удвоенный отступ здесь и не всё в порядке с отступами в заголовках у трёх функций ниже

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+auto HashMap<KeyType, ValueType, Hash>::find(const KeyType key) -> iterator {
+    size_t hashFunc = function_hash_number_(key) %
+                      hash_table_size_quantity_elements_;
+    for (auto& It : hash_table_[hashFunc]) {
```
hashFunc -> hash_table_cell

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+template <class KeyType, class ValueType, class Hash>
+void HashMap<KeyType, ValueType, Hash>::insert(
+    std::pair<KeyType, ValueType> key_value_pair) {
+    size_t hash_table_cell_id = function_hash_number_(key_value_pair.first) %
```
1) лучше hash_table_cell, как вы везде назвали
2) более читаемо получается, если операторы на новой строчке:
```
int value = (very_very_long_expression)
    % (another_very_very_long_expression);
```
Исправьте здесь и во всех аналогичных местах.

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,272 @@
+template <class KeyType, class ValueType, class Hash>
+void HashMap<KeyType, ValueType, Hash>::clear() {
+    for (auto& element : all_elements_) {
+        size_t hashFunction = function_hash_number_(element.first) %
```
hashFunction -> hash_table_cell

SenderOK reviewed (changes requested):
Исправляйте всё, перепосылайте в Яндекс.Контест (не забывайте комментарием дать ссылку на посылку) и перезапрашивайте ревью

IlyaChukanov committed:
Update hash_map.h

https://official.contest.yandex.ru/contest/17022/problems/?success=31742653#39576/2016_12_05/m50MX2nMXF

IlyaChukanov committed:
Update hash_map.h

https://official.contest.yandex.ru/contest/17022/problems/?success=31742972#39576/2016_12_05/m50MX2nMXF

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+template <class KeyType, class ValueType, class Hash>
+HashMap<KeyType, ValueType, Hash>& HashMap<KeyType, ValueType, Hash>::operator=(
+        HashMap<KeyType, ValueType, Hash> const& other) {
+    auto It = this;
```
эта переменная лишняя, лучше делать `return *this`

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+template <class KeyType, class ValueType, class Hash = std::hash<KeyType>>
+class HashMap {
+    using KeyValuePair = std::pair<const KeyType, ValueType>;
+    using KeyValuePair2 = std::pair<KeyType, ValueType>;
```
кажется, можно KeyValuePair2 убрать, везде использовать KeyValuePair

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+auto HashMap<KeyType, ValueType, Hash>::find(KeyType key) const
+        -> const_iterator {
+    size_t hash_table_cell = hash_function_(key) % hash_table_size_;
+    for (auto& It : hash_table_[hash_table_cell]) {
```
It -> it, нужно везде исправить

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+
+    HashMap& operator=(HashMap const& other);
+
+    iterator find(const KeyType key);
```
здесь и в нескольких методах ниже `const KeyType key` -> `const KeyType& key`

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+
+template <class KeyType, class ValueType, class Hash>
+HashMap<KeyType, ValueType, Hash>::HashMap(
+    std::initializer_list<KeyValuePair2> list, Hash hasher_obj)
```
если уже даже один параметр не влазит на строку, то лучше каждый на своей строке с удвоенным отступом (в вашем случае 8): https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+
+template <class KeyType, class ValueType, class Hash>
+auto HashMap<KeyType, ValueType, Hash>::end() const -> const_iterator {
+    return all_elements_.end();
```
.cend()

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+
+template <class KeyType, class ValueType, class Hash>
+auto HashMap<KeyType, ValueType, Hash>::begin() const -> const_iterator {
+    return all_elements_.begin();
```
.cbegin()

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,265 @@
+
+template <class KeyType, class ValueType, class Hash>
+auto HashMap<KeyType, ValueType, Hash>::end() -> iterator {
+    return iterator(all_elements_.end());
```
кажется, здесь и в симметричной функции не нужно вот это оборачивание в iterator()

SenderOK reviewed (changes requested):
правьте, прикладывайте посылку в Контест, перезапрашивайте ревью

IlyaChukanov committed:
Update hash_map.h

https://official.contest.yandex.ru/contest/17022/run-report/31804193/

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,264 @@
+
+    const ValueType& at(const KeyType key) const;
+
+    const_iterator find(const KeyType key) const;
```
в этих трёх функциях забыли &

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,264 @@
+
+template <class KeyType, class ValueType, class Hash>
+HashMap<KeyType, ValueType, Hash>::HashMap(
+        std::initializer_list<KeyValuePair> list, Hash hasher_obj)
```
Hash hasher_obj на следующую строку

SenderOK commented on hash_map.h:
```diff
@@ -0,0 +1,264 @@
+
+    bool empty() const;
+
+    void insert(KeyValuePair key_value_pair);  // тут KeyValue не const
```
1) `const KeyValuePair key_value_pair&`
2) уберите этот коммент тут и выше

SenderOK reviewed (approved):
правьте последние замечания, прогоняйте в Контесте и вливайте

IlyaChukanov committed:
Update hash_map.h

https://official.contest.yandex.ru/contest/17022/run-report/31814826/
